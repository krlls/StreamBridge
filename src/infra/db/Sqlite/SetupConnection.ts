import { DataSource, EntityTarget, Repository } from 'typeorm'
import Database from 'better-sqlite3'

import { NODE_ENV, sqlLite } from '../../../config'

export const dataSource = new DataSource(sqlLite[NODE_ENV])

export class SqliteDB {
  private static _instance: SqliteDB

  public static get instance(): SqliteDB {
    if (!this._instance) this._instance = new SqliteDB()

    return this._instance
  }

  private dbConnect!: DataSource
  private testdb!: any

  async setupTestDB() {
    this.testdb = new Database('')
    this.dbConnect = await dataSource.initialize()
  }

  async teardownTestDB() {
    await this.dbConnect.destroy()
    this.testdb.close()
  }
}

export const getRepository = <Entity extends Record<string, any>>(target: EntityTarget<Entity>): Repository<Entity> => {
  return dataSource.getRepository(target)
}

export default dataSource
