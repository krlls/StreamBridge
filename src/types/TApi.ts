export namespace Api {
  export namespace User {
    export const PREFIX = '/user'

    export namespace Create {
      export const URL = '/create'

      export type Req = {
        login: string,
        pass: string,
      }
    }
  }
}
