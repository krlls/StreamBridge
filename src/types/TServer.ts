export type ServerConfig = {
  port: number,
  silent: boolean,
}
