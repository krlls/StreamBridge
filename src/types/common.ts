export interface Converter<F, T> {
  from(from: F): T,
  to(to: T): F,
}

export enum E_NODE_ENV {
  TEST = 'test',
  DEV = 'development',
  PROD = 'production',
}
