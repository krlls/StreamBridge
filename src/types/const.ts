const TYPES = {
  UserController: Symbol.for('UserController'),

  UserService: Symbol.for('UserService'),

  UserRepository: Symbol.for('UserRepository'),

  UserEntityConverter: Symbol.for('UserEntityConverter'),
}

export { TYPES }
