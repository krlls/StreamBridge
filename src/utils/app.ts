import { isUndefined } from 'lodash'

export const requiredEnv = <T extends string | number>(env?: T, name?: string) => {
  if (isUndefined(env)) {
    throw Error(`Required env ${name} not set`)
  }

  return env
}
