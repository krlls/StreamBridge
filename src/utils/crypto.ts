import * as argon2 from 'argon2'

export const hashPass = async (pass: string) => argon2.hash(pass)
export const verifyPass = async (pass: string, hash: string) => argon2.verify(hash, pass)
