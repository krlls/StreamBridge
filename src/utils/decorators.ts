import { get, isFunction } from 'lodash'

export const strategy =
  <T extends { prototype: any }>(checkProp: string, method: string) =>
  ({ prototype }: T) => {
    const propertyDescriptors = Object.getOwnPropertyDescriptors(prototype)

    for (const propertyName in propertyDescriptors) {
      const descriptor = propertyDescriptors[propertyName]

      if (!isFunction(descriptor.value) || ['constructor', method].includes(propertyName)) {
        continue
      }

      const originalMethod = descriptor.value

      descriptor.value = function <T extends []>(...args: T) {
        if (!get(this, checkProp)) {
          throw new Error(`You must call the ${method}()`)
        }

        return originalMethod.apply(this, args)
      }

      Object.defineProperty(prototype, propertyName, descriptor)
    }
  }
