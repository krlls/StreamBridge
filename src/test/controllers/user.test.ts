import { expect, test, describe, beforeEach, afterEach } from '@jest/globals'

import { Api } from '../../types/TApi'
import { createPatch } from '../../utils/links'
import { TestApp } from '../index.test'
import { SqliteDB } from '../../infra/db/Sqlite/SetupConnection'

describe('User tests', () => {
  beforeEach(async () => {
    await SqliteDB.instance.setupTestDB()
  })

  afterEach(async () => {
    await SqliteDB.instance.teardownTestDB()
  })

  const vkPatch: (...args: string[]) => string = createPatch.bind(null, Api.User.PREFIX)

  test('Create user works', async () => {
    const validUserData = {
      login: 'Ksmi',
      pass: '123',
    }

    const response = await TestApp.post(vkPatch(Api.User.Create.URL)).send(validUserData)

    expect(response.status).toBe(200)
    expect(response.text).toMatchSnapshot()
  })

  test('Create user with not valid', async () => {
    const notValidUserData = {
      login: 1234,
      pass: 123,
    }

    const response = await TestApp.post(vkPatch(Api.User.Create.URL)).send(notValidUserData)

    expect(response.status).toBe(400)
    expect(response.text).toMatchSnapshot()
  })
})
