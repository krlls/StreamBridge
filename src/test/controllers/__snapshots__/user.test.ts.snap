// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User tests Create user with not valid 1`] = `"{"message":"Validation error","details":[{"message":"\\"login\\" must be a string","path":["login"],"type":"string.base","context":{"label":"login","value":1234,"key":"login"}}]}"`;

exports[`User tests Create user works 1`] = `"{"id":1,"login":"Ksmi"}"`;
